File: frontend/src/utils/permissions.js
// 前端权限工具：基于后端 /auth/me 返回的权限矩阵驱动 UI 隐藏/禁用
//
// 后端 require_permission 是真正鉴权防线，前端只做 UI 隐藏/禁用以提升体验。
// 权限矩阵来源：登录/刷新时 /auth/me 返回的 user.permissions（与后端 Role.permissions 一致）。
// admin 角色自动拥有全部权限（后端 _resolve_permissions 已展开为全量矩阵）。

import { useAuthStore } from '../store/authStore'

/**
 * 判断当前用户是否拥有指定模块动作权限（前端仅用于 UI 隐藏）。
 * @param {string} module 模块名，如 "user" / "workflow"
 * @param {string} action 动作名，如 "view" / "edit" / "delete"
 * @returns {boolean}
 */
export function hasPermission(module, action) {
  const user = useAuthStore.getState().user
  if (!user) return false
  // admin 角色自动通过（后端已返回全量权限矩阵，此为兜底）
  if (user.role === 'admin') return true
  // 优先使用后端返回的权限矩阵（支持自定义角色）
  const perms = user.permissions
  if (perms && typeof perms === 'object') {
    const actions = perms[module]
    return Array.isArray(actions) && actions.includes(action)
  }
  return false
}

/**
 * 是否为管理员（admin 角色）。
 */
export function isAdmin() {
  const user = useAuthStore.getState().user
  return !!user && user.role === 'admin'
}

/**
 * 判断当前用户是否可编辑指定资源（资源级 owner 控制）。
 *
 * 优先级：
 * 1. admin 角色可编辑所有资源；
 * 2. 后端返回的 ``can_edit`` 标志（列表/详情接口已计算，覆盖 owner 与被授权用户）；
 * 3. 兜底：当前用户是资源创建者（``created_by === user.id``），用于详情接口未返回 can_edit 时。
 *
 * @param {object|null} resource 资源对象（列表项或详情），需含 can_edit 或 created_by 字段
 * @returns {boolean}
 */
export function canEditResource(resource) {
  const user = useAuthStore.getState().user
  if (!user) return false
  if (user.role === 'admin') return true
  if (resource && typeof resource.can_edit === 'boolean') return resource.can_edit
  // 兜底：详情接口若未返回 can_edit，则按 owner 判断（被授权用户需依赖 can_edit）
  if (resource && resource.created_by != null) return resource.created_by === user.id
  return false
}

/**
 * 判断当前用户是否可管理资源共享（仅 admin 或 owner）。
 *
 * 被授权用户虽可编辑资源，但不能管理共享（不能再转授或撤销）。
 * @param {object|null} resource 资源对象，需含 created_by 字段
 * @returns {boolean}
 */
export function canManageShare(resource) {
  const user = useAuthStore.getState().user
  if (!user) return false
  if (user.role === 'admin') return true
  if (resource && resource.created_by != null) return resource.created_by === user.id
  return false
}

export default { hasPermission, isAdmin, canEditResource, canManageShare }

File: backend/app/core/permissions.py
"""权限矩阵定义。

按模块划分权限，每个模块包含若干动作（view/edit/delete/execute 等）。
角色通过 permissions JSON 字段存储 ``{模块: [动作列表]}`` 结构。

用于 ``require_permission(module, action)`` 依赖校验。

模块粒度与侧边栏目录保持一致：
- 工作流拆分为「工作流列表」「工作流编排」两个子模块；
- 资产管理拆分为「资产总览」「资产清单」「类型模板」三个子模块；
- 对话、设备对接从 agent / llm_config 中独立为单独模块。
"""
from typing import Any

# 权限模块与可用动作
# 新增模块时在此注册，前端角色管理页面会动态读取
PERMISSION_MODULES: dict[str, list[str]] = {
    # 运营中心
    "dashboard": ["view"],
    "approval": ["view", "approve"],
    "chat": ["view"],
    # 智能体编排
    "agent": ["view", "edit", "delete", "execute"],
    "workflow_list": ["view", "edit", "delete", "execute"],
    "workflow_editor": ["view", "edit"],
    "skill": ["view", "edit", "delete"],
    "tool": ["view", "edit", "delete", "execute"],
    "knowledge_base": ["view", "edit", "delete"],
    # 资源管理
    "asset_overview": ["view"],
    "asset_list": ["view", "edit", "delete"],
    "asset_templates": ["view", "edit", "delete"],
    "deliverable": ["view", "edit", "delete"],
    # 连接配置
    "llm_config": ["view", "edit", "delete"],
    "device": ["view", "edit"],
    # 运行监控
    "execution": ["view", "export"],
    "notification": ["view", "edit"],
    # 告警接入（解析策略管理 / 告警列表 / 入库监控）
    "strategy": ["view", "edit", "delete"],
    "alert": ["view", "export"],
    "monitor": ["view", "edit"],
    # 告警自动封禁工作流（实例管理 / 封禁审批 / 封禁工作台 / 触发规则）
    "ban_workflow": ["view", "edit"],
    # 系统管理
    "user": ["view", "edit", "delete"],
    "role": ["view", "edit", "delete"],
    "feedback": ["view", "edit", "delete"],
    "system_config": ["view", "edit"],
    "system_monitor": ["view"],
    "audit_log": ["view"],
    # 运营管理（值班管理拆分为二级权限：人员/值班表/请假/调班/大屏）
    "duty_member": ["view", "edit", "delete"],
    "duty_schedule": ["view", "edit", "delete"],
    "duty_leave": ["view", "edit", "delete"],
    "duty_log": ["view", "edit", "delete"],
    "duty_dashboard": ["view"],
}

# 模块中文标签（前端展示用）
MODULE_LABELS: dict[str, str] = {
    # 运营中心
    "dashboard": "运营大屏",
    "approval": "工作台",
    "chat": "对话",
    # 智能体编排
    "agent": "智能体",
    "workflow_list": "工作流列表",
    "workflow_editor": "工作流编排",
    "skill": "技能",
    "tool": "工具",
    "knowledge_base": "知识库",
    # 资源管理
    "asset_overview": "资产总览",
    "asset_list": "资产清单",
    "asset_templates": "类型模板",
    "deliverable": "材料管理",
    # 连接配置
    "llm_config": "模型设置",
    "device": "设备对接",
    # 运行监控
    "execution": "日志中心",
    "notification": "通知中心",
    # 告警接入（解析策略管理 / 告警列表 / 入库监控）
    "strategy": "解析策略",
    "alert": "告警列表",
    "monitor": "入库监控",
    "ban_workflow": "封禁工作流",
    # 系统管理
    "user": "用户管理",
    "role": "角色管理",
    "feedback": "反馈管理",
    "system_config": "系统设置",
    "system_monitor": "系统监控",
    "audit_log": "审计日志",
    # 运营管理（值班管理二级权限）
    "duty_member": "值班人员",
    "duty_schedule": "值班表",
    "duty_leave": "请假管理",
    "duty_log": "调班记录",
    "duty_dashboard": "值班大屏",
}

# 动作中文标签
ACTION_LABELS: dict[str, str] = {
    "view": "查看",
    "edit": "编辑",
    "delete": "删除",
    "execute": "执行",
    "approve": "审批",
    "export": "导出",
}

# 权限模块迁移映射（旧模块 → 新模块子项）
# 用于把旧角色 JSON 中的 workflow / asset 权限展开到细分模块
_LEGACY_MODULE_MIGRATION: dict[str, list[str]] = {
    "workflow": ["workflow_list", "workflow_editor"],
    "asset": ["asset_overview", "asset_list", "asset_templates"],
    "duty": ["duty_member", "duty_schedule", "duty_leave", "duty_log", "duty_dashboard"],
}


def migrate_permissions(perms: dict[str, list[str]] | None) -> dict[str, list[str]]:
    """把旧版权限矩阵迁移到新结构（保留已有动作，去重）。

    - ``workflow`` 的动作 → 复制到 ``workflow_list``（全量）+ ``workflow_editor``（仅 view/edit）
    - ``asset`` 的动作 → 复制到 ``asset_overview``（仅 view）+ ``asset_list`` + ``asset_templates``
    - 旧键保留，便于回滚；新键已存在则不覆盖
    - 兼容性补齐：原 ``chat`` 由 ``agent`` 权限覆盖、原 ``device`` 由 ``llm_config`` 覆盖；
      迁移时为非空角色补 ``chat: ['view']`` 和 ``device: ['view']``，避免现有角色看不到菜单
    """
    if not perms:
        return {}
    result = {k: list(v) for k, v in perms.items()}

    for old_mod, new_mods in _LEGACY_MODULE_MIGRATION.items():
        old_actions = perms.get(old_mod)
        if not old_actions:
            continue
        for new_mod in new_mods:
            if new_mod in result and result[new_mod]:
                continue  # 已显式配置，不覆盖
            # 子模块按其动作集 ⨯ 旧动作取交集
            allowed = PERMISSION_MODULES.get(new_mod, [])
            inherited = [a for a in old_actions if a in allowed]
            if inherited:
                result[new_mod] = inherited

    # 兼容性补齐：chat / device 从原 agent / llm_config 中独立出来
    # 原有 agent 权限覆盖对话 → 自动补 chat:view
    if "chat" not in result and perms.get("agent"):
        result["chat"] = ["view"]
    # 原有 llm_config 权限覆盖设备对接 → 自动补 device:view（仅 view，edit 需手动分配）
    if "device" not in result and perms.get("llm_config"):
        result["device"] = ["view"]

    return result


# 默认角色权限（种子数据用）
DEFAULT_ROLES: list[dict[str, Any]] = [
    {
        "name": "admin",
        "description": "超级管理员，拥有所有权限",
        "is_system": True,
        "permissions": {mod: list(actions) for mod, actions in PERMISSION_MODULES.items()},
    },
    {
        "name": "analyst",
        "description": "安全运营人员，可编辑工作流/智能体/知识库、处理审批",
        "is_system": True,
        "permissions": {
            "dashboard": ["view"],
            "approval": ["view", "approve"],
            "chat": ["view"],
            "agent": ["view", "edit", "execute"],
            "workflow_list": ["view", "edit", "delete", "execute"],
            "workflow_editor": ["view", "edit"],
            "skill": ["view", "edit"],
            "tool": ["view", "execute"],
            "knowledge_base": ["view", "edit"],
            "llm_config": ["view"],
            "device": ["view"],
            "asset_overview": ["view"],
            "asset_list": ["view", "edit", "delete"],
            "asset_templates": ["view", "edit", "delete"],
            "deliverable": ["view", "edit"],
            "duty_member": ["view", "edit", "delete"],
            "duty_schedule": ["view", "edit", "delete"],
            "duty_leave": ["view", "edit", "delete"],
            "duty_log": ["view", "edit", "delete"],
            "duty_dashboard": ["view"],
            "execution": ["view", "export"],
            "notification": ["view"],
            "strategy": ["view"],
            "ban_workflow": ["view"],
            "feedback": ["view"],
            "system_monitor": ["view"],
            "audit_log": ["view"],
        },
    },
    {
        "name": "viewer",
        "description": "只读用户，仅可查看",
        "is_system": True,
        "permissions": {
            "dashboard": ["view"],
            "approval": ["view"],
            "chat": ["view"],
            "agent": ["view"],
            "workflow_list": ["view"],
            "workflow_editor": ["view"],
            "skill": ["view"],
            "tool": ["view"],
            "knowledge_base": ["view"],
            "llm_config": ["view"],
            "device": ["view"],
            "asset_overview": ["view"],
            "asset_list": ["view"],
            "asset_templates": ["view"],
            "deliverable": ["view"],
            "duty_member": ["view"],
            "duty_schedule": ["view"],
            "duty_leave": ["view"],
            "duty_log": ["view"],
            "duty_dashboard": ["view"],
            "execution": ["view"],
            "notification": ["view"],
            "strategy": ["view"],
            "alert": ["view"],
            "monitor": ["view"],
            "ban_workflow": ["view"],
            "feedback": ["view"],
            "system_monitor": ["view"],
            "audit_log": ["view"],
        },
    },
]


def has_permission(permissions: dict[str, list[str]] | None, module: str, action: str) -> bool:
    """检查权限矩阵是否包含指定模块的动作。

    Args:
        permissions: 角色的 permissions 字典 ``{模块: [动作]}``
        module: 权限模块名
        action: 动作名

    Returns:
        是否有权限
    """
    if not permissions:
        return False
    actions = permissions.get(module, [])
    return action in actions
